File: javascript/valid-anagram.js
/*
 * @lc app=leetcode id=242 lang=javascript
 *
 * [242] Valid Anagram
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
    if (s.length != t.length) return false;
    sMap = new Map();
    tMap = new Map();
    for (let i = 0; i < s.length; i++) {
        sMap.set(s[i], (sMap.get(s[i]) || 1) + 1);
        tMap.set(t[i], (tMap.get(t[i]) || 1) + 1);
    }
    for (const [char, count] of sMap) {
        if (sMap.get(char) != tMap.get(char)) return false;
    }
    return true;
};
// @lc code=end

/*
Hash Maps
Time Complexity: O(n + m)
Space Complexity: O(1) - 26 alphabets
*/

File: python3/valid-anagram.py
#
# @lc app=leetcode id=242 lang=python3
#
# [242] Valid Anagram
#

# @lc code=start
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): 
            return False
        
        s_char_dict = dict()
        t_char_dict = dict()

        for i in range(len(s)):
            s_char_dict[s[i]] = s_char_dict.get(s[i], 0) + 1
            t_char_dict[t[i]] = t_char_dict.get(t[i], 0) + 1
        
        return s_char_dict == t_char_dict
                
# @lc code=end

